---
title: "Using SynExtend - Simple"
excerpt: "A simple workflow with SynExtend, simply using a synteny map to perform orthology inference between two genomes."
collection: portfolio
output:
  md_document:
    variant: gfm
    preserve_yaml: TRUE
    toc: TRUE
    toc_depth: 3
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile,
    encoding = encoding,
    output_dir = "../_portfolio") })
always_allow_html: true
---

# header 1

```{r setup, include = FALSE}

# base_dir <- "~/Dropbox/svmiller.github.io/" # i.e. where the jekyll blog is on the hard drive.
# base_url <- "/" # keep as is
# fig_path <- "images/" # customize to heart's content, I 'spose.
# 
# knitr::opts_knit$set(base.dir = base_dir, base.url = base_url)
# knitr::opts_chunk$set(fig.path = fig_path,
#                       cache.path = '../cache/',
#                       message=FALSE, warning=FALSE,
#                       cache = TRUE) 
suppressMessages(library(knitr))
base_dir <- "~/Repos/Personal_Website/npcooley.github.io/"
base_url <- "/"
fig_path <- "images/"
opts_knit$set(base.dir = base_dir,
              base.url = base_url)
opts_chunk$set(fig.path = fig_path,
               cache.path = "../cache/",
               message = FALSE,
               warning = FALSE,
               cache = FALSE)
```

```{r, checkthings}
list.files("../images")
```

## subheader 1

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

# header 2

## Including Plots

### minor header

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

```{r stack, echo = FALSE}
plot(stackloss$Water.Temp)
```


```{r dirstuff, echo = FALSE, include = FALSE}
# leaving this in for now as breadcrumb for when i need to come back to it
list.files("images",
           full.names = TRUE)
list.files("../images")
getwd()
# setting cache to false seems to make this unnecessary, but there's still a lot happening here under the hood that i don't seem to understand
# file.copy(from = list.files("images",
#                             full.names = TRUE),
#           to = "../images/")
# file.remove(list.files("images",
#                             full.names = TRUE))
# file.remove("images")
```

# header 3

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
